### PR TITLE
EHN: Add support for vector atol and rtol in np.all_isclose and np…

### LINKED ISSUE #14320: allclose() vector `atol` parameter wrong shape if NaN in a or b.
```python
>>>np.allclose([1,2,float('nan')], [1,2,float('nan')], equal_nan=True)
True
>>>np.allclose([1,2,float('nan')], [1,2,float('nan')], atol=[1,1], equal_nan=True)
True
>>>np.allclose([1,2,3], [1,2,3)], atol=[1,1,1], equal_nan=True)
True
>>>np.allclose([1,2,float('nan')], [1,2,float('nan')], atol=[1,1,1], equal_nan=True)
ValueError: operands could not be broadcast together with shapes (3,) (2,)
```

It appears that `allclose` is basically removing entries where `a` or `b` is NaN before passing them on to `isclose` with the same `atol` provided to `allclose`. This of course doesn't work if I expect there to be no NaNs some of the time and need to provide a longer `atol` because my `a` and `b` are different scales in different positions.

What should happen is `allclose` removes the same positions from `atol` before passing it on, so the dimensions always agree.

### PR DESCRIPTION
**Pull request by maxwell-aladago**
Fixes #14320.

### CODE REVIEW

**Comment by charris** [comment]
Fixes #14320.

**Comment by charris** [comment]
Needs a release note.

**Comment by maxwell-aladago** [comment]
@charris, will this be ok for a release note?

```
Vector ``atol`` and ``rtol`` in ``np.all_isclose`` and ``np.isclose``
-----------------------------------------------------------------------
The keywords ``atol`` and ``rtol`` in ``np.isclose`` and ``np.all_isclose`` 
now accept both scalars and vectors. A vector if given must have the same 
length as the larger of the two input arrays ``x`` and ``y``. 
```

**Comment by maxwell-aladago** [comment]
By the way, the component is `numpy.core`, not `numpy.testing`

File: doc/release/upcoming_changes/14343.future_change.rst [hunk 1]
```diff
@@ -0,0 +1,5 @@
+Vector ``atol`` and ``rtol`` in ``np.all_isclose`` and ``np.isclose``
```
Comment (eric-wieser) [inline doc/release/upcoming_changes/14343.future_change.rst]:
```suggestion
``np.all_isclose`` and ``np.isclose`` broadcast their ``atol`` and ``rtol`` arguments
```

File: doc/release/upcoming_changes/14343.future_change.rst [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+-----------------------------------------------------------------------
+The keywords ``atol`` and ``rtol`` in ``np.isclose`` and ``np.all_isclose``
+now accept both scalars and vectors. A vector if given must have the same
+length as the larger of the two input arrays ``x`` and ``y``.
```
Comment (eric-wieser) [inline doc/release/upcoming_changes/14343.future_change.rst]: This isn't really accurate. Better would be "accepts both scalars and arrays", and "An array, if given, must broadcast to the shapes of `x` and `y`."

File: numpy/core/numeric.py [hunk 3]
```diff
@@ -2250,8 +2250,10 @@ def within_tol(x, y, atol, rtol):
         # lib.stride_tricks, though, so we can't import it here.
         x = x * ones_like(cond)
         y = y * ones_like(cond)
+        rtol = full(x.shape, rtol)
```
Comment (shoyer) [inline numpy/core/numeric.py]: Use np.broadcast_to

**Comment by mattip** [comment]
@eric-wieser, @shoyer any more comments?

**Comment by charris** [comment]
close/reopen

**Comment by mattip** [comment]
Is there a clear use case for this? I feel it is a bit too complicated to have non-scalar atol/rtol.

**Comment by mattip** [comment]
If you disagree feel free to reopen. Please also describe the use case in the mailing list since this is an enhancement request.

**Comment by mattip** [comment]
Could you write to [numpy-discussion](https://mail.python.org/mailman3/lists/numpy-discussion.python.org/) proposing the enhancement?

**Comment by seberg** [comment]
Anyone want to finish this up?  A new PR would also be fine.  The docs are slightly off, there is no `all_isclose` function and since I doubt it is super common, maybe we should keep the scalar case and only do `?tol[finite]` (i.e. the indexing) when it isn't scalar?

**Comment by mattip** [comment]
Closing, #24878 is a duplicate and was merged. Thanks @maxwell-aladago anyway.

**Pull request closed without merging**